File: python3/reverse_linked_list_ii.py
from model import ListNode
from typing import Optional

# Space : O(n)
# Time  : O(n)


class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if left == right:
            return head

        ans = ListNode(0)
        ans.next = head
        pre = ans

        for i in range(left - 1):
            pre = pre.next
        
        # reverse the [left, right] nodes
        reverse = None
        cur = pre.next
        for i in range(right - left + 1):
            next = cur.next
            cur.next = reverse
            reverse = cur
            cur = next

        pre.next.next = cur
        pre.next = reverse

        return ans.next

File: golang/reverse_linked_list_ii.go
package main

import "golang/model"

func reverseBetween(head *model.ListNode, left int, right int) *model.ListNode {
	if left == right {
		return head
	}

	var temp []int
	p := head
	for p != nil {
		temp = append(temp, p.Val)
		p = p.Next
	}

	for left < right {
		temp[left-1], temp[right-1] = temp[right-1], temp[left-1]
		left++
		right--
	}

	new_head := &model.ListNode{
		Val: temp[0],
	}
	p = new_head
	for i := 1; i < len(temp); i++ {
		p.Next = &model.ListNode{
			Val: temp[i],
		}
		p = p.Next
	}

	return new_head
}